### PR TITLE
[4.2] Convert basic API auth plugin to services

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Converts the basic API auth plugin to services.

### Testing Instructions
- Enable the basic API auth plugin
- Make an API request with basic authentication, for example you can use curl or postman

### Actual result BEFORE applying this Pull Request
All works.

### Expected result AFTER applying this Pull Request
All works.

### CODE REVIEW

File: plugins/api-authentication/basic/basic.xml [hunk 1]
```diff
@@ -9,8 +9,10 @@
 	<authorUrl>www.joomla.org</authorUrl>
 	<version>4.0.0</version>
 	<description>PLG_API-AUTHENTICATION_BASIC_XML_DESCRIPTION</description>
+	<namespace path="src">Joomla\Plugin\ApiAuthentication\Basic</namespace>
 	<files>
-		<filename plugin="basic">basic.php</filename>
+		<folder plugin="cache">services</folder>
```
Comment (heelc29) [inline plugins/api-authentication/basic/basic.xml]:
```suggestion
		<folder plugin="basic">services</folder>
```

File: plugins/api-authentication/basic/services/provider.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	 */
+	public function register(Container $container)
+	{
+		$container->set(
+			PluginInterface::class,
+			function (Container $container)
+			{
+				$plugin = new Basic(
+					$container->get(DispatcherInterface::class),
+					(array) PluginHelper::getPlugin('system', 'cache'),
```
Comment (heelc29) [inline plugins/api-authentication/basic/services/provider.php]:
```suggestion
					(array) PluginHelper::getPlugin('api-authentication', 'basic'),
```

**Comment by heelc29** [comment]
Can you please also correct the filenames of the language files in the manifest to: `plg_api-authentication_basic(.sys).ini` (wrong added in #23424 😅)

https://github.com/joomla/joomla-cms/blob/ee81859803672f1c6fee40a1553be3ae7fb47485/plugins/api-authentication/basic/basic.xml#L16-L17

**Comment by laoneo** [comment]
Done

File: plugins/api-authentication/basic/basic.xml [hunk 3]
```diff
@@ -9,11 +9,13 @@
...
 	<files>
-		<filename plugin="basic">basic.php</filename>
+		<folder plugin="basic">services</folder>
+		<folder>src</folder>
 	</files>
 	<languages>
-		<language tag="en-GB">language/en-GB/plg_authentication_api_basic.ini</language>
-		<language tag="en-GB">language/en-GB/plg_authentication_api_basic.sys.ini</language>
+		<language tag="en-GB">language/en-GB/plg_authentication-api_basic.ini</language>
+		<language tag="en-GB">language/en-GB/plg_authentication-api_basic.sys.ini</language>
```
Comment (heelc29) [inline plugins/api-authentication/basic/basic.xml]:
```suggestion
		<language tag="en-GB">language/en-GB/plg_api-authentication_basic.ini</language>
		<language tag="en-GB">language/en-GB/plg_api-authentication_basic.sys.ini</language>
```
api before authentication (see administrator/language/en-GB/plg_api-authentication_basic.ini)

File: plugins/api-authentication/basic/services/provider.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
+<?php
+/**
+ * @package     Joomla.Plugin
+ * @subpackage  Authentication.joomla
```
Comment (heelc29) [inline plugins/api-authentication/basic/services/provider.php]:
```suggestion
 * @subpackage  Apiauthentication.basic
```
Also in Basic.php

File: plugins/api-authentication/basic/basic.xml (hunk 3 above)
Comment (laoneo) [reply plugins/api-authentication/basic/basic.xml]: :see_no_evil:

File: plugins/api-authentication/basic/services/provider.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+use Joomla\CMS\Plugin\PluginHelper;
+use Joomla\CMS\User\UserFactoryInterface;
+use Joomla\Database\DatabaseInterface;
+use Joomla\DI\Container;
+use Joomla\DI\ServiceProviderInterface;
+use Joomla\Event\DispatcherInterface;
+use Joomla\Plugin\ApiAuthentication\Basic\Basic;
+
+return new class implements ServiceProviderInterface {
+	/**
```
Comment (heelc29) [inline plugins/api-authentication/basic/services/provider.php]:
```suggestion
return new class implements ServiceProviderInterface
{
	/**
```

File: plugins/api-authentication/basic/src/Basic.php [hunk 6]
```diff
@@ -1,42 +1,65 @@
...
 
 	/**
-	 * The application object
+	 * The user factory
 	 *
-	 * @var    \Joomla\Database\DatabaseInterface
-	 * @since  4.0.0
+	 * @var    UserFactoryInterface
+	 *
+	 * @since  __DEPLOY_VERSION__
```
Comment (heelc29) [inline plugins/api-authentication/basic/src/Basic.php]:
```suggestion
	 * @var    UserFactoryInterface
	 * @since  __DEPLOY_VERSION__
```

**Review by heelc29 — CHANGES_REQUESTED** [review]
cs

**Comment by heelc29** [comment]
In other extensions the file in the `src` folder is in a subfolder `Extension`. Should it also?

**Comment by heelc29** [comment]
I have tested this item :white_check_mark: successfully on 94bbd50c0916719f101192a0572e0810abdc2897<br /><br />Thanks @laoneo<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37803">issues.joomla.org/tracker/joomla-cms/37803</a>.</sub>

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 94bbd50c0916719f101192a0572e0810abdc2897<br /><br />Tested successfully in Joomla 4.2.0-alpha3-dev of 21 May in Wampserver 3.2.8 using PHP 8.0.15.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37803">issues.joomla.org/tracker/joomla-cms/37803</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37803">issues.joomla.org/tracker/joomla-cms/37803</a>.</sub>

**Comment by roland-d** [comment]
Thanks everybody

**Pull request merged**